### PR TITLE
feat: Multiple Headed DoConcurrentLoop Total Iterations

### PR DESCRIPTION
**Pull request by adit4443ya**
Towards #4900

Intended towards correctly calculating total number of iterations in Multiple Headed DoConcurrentLoop.

### CODE REVIEW

File: integration_tests/openmp_40.f90 [hunk 1]
```diff
@@ -0,0 +1,6 @@
+program openmp_40
+integer :: i, j
+do concurrent ( i =1:4, j = 1:3 )
+    print *,i
+end do 
+end program
```
Comment (certik) [inline integration_tests/openmp_40.f90:6]: I would write this test in some way to ensure it actually ran correctly. For example assigning to some array, and then checking that all elements were correctly assigned, or doing some reduction and checking the final sum.

**Comment by adit4443ya** [comment]
For correctly calculating the iterating variables 
I have thought of something like this 
```diff
(lf) aditya-trivedi@Observer:~/thats_me/main/lfortran$ git diff
diff --git a/src/libasr/pass/openmp.cpp b/src/libasr/pass/openmp.cpp
index 457673612..761c0a239 100644
--- a/src/libasr/pass/openmp.cpp
+++ b/src/libasr/pass/openmp.cpp
@@ -932,11 +932,51 @@ class DoConcurrentVisitor :
                 }
             }
 
-            std::vector<ASR::stmt_t*> loop_body;
-            for (size_t i = 0; i < do_loop.n_body; i++) {
-                loop_body.push_back(do_loop.m_body[i]);
+            // integer :: I = 0;
+            std::vector<ASR::stmt_t*> flattened_body;
+            ASR::expr_t* I = b.Variable(current_scope, "I", ASRUtils::TYPE(ASR::make_Integer_t(al, loc,
+            4)),ASR::intentType::Local, ASR::abiType::BindC);
+
+            ASR::expr_t* temp_I = I;
+            for (size_t i = 0; i < do_loop.n_head; ++i) {
+                ASR::do_loop_head_t head = do_loop.m_head[i];
+                ASR::expr_t* computed_var;
+                std::cout<<"Hello Flattened Body!"<<std::endl;
+
+                if (i == do_loop.n_head - 1) {
+                    // Last loop variable -> ik = (I % (nk - ak + 1)) + ak
+                    Vec<ASR::expr_t*> mod_args; mod_args.reserve(al, 2);
+                    mod_args.push_back(al, temp_I);
+                    mod_args.push_back(al, dimension_lengths[i]);
+                    computed_var = b.Add(ASRUtils::EXPR(ASRUtils::make_IntrinsicElementalFunction_t_util(al,
+                    loc,2,mod_args.p, 2, 0, ASRUtils::expr_type(dimension_lengths[i]), nullptr)),head.m_start);
+                    // computed_var = b.Add(b.Mod(temp_I, dimension_lengths[i]), head.m_start);
+                } else {
+                        // Intermediate loop variable -> iy = ((I / ((nz - az + 1) * (nk - ak + 1))) % (ny - ay +1)) + ay 
+                        ASR::expr_t* product_of_next_dimensions = b.i32(1); for (size_t j = i + 1; j <
+                        do_loop.n_head; ++j) {
+                            product_of_next_dimensions = b.Mul(product_of_next_dimensions, dimension_lengths[j]);
+                    }
+                    Vec<ASR::expr_t*> mod_args; mod_args.reserve(al, 2);
+                    mod_args.push_back(al, b.Div(temp_I, product_of_next_dimensions));
+                    mod_args.push_back(al, dimension_lengths[i]);
+                    computed_var = b.Add(ASRUtils::EXPR(ASRUtils::make_IntrinsicElementalFunction_t_util(al,
+                    loc, 2,mod_args.p, 2, 0, ASRUtils::expr_type(dimension_lengths[i]), nullptr)),head.m_start);
+                    // computed_var = b.Add(b.Mod(b.Div(temp_I, product_of_next_dimensions),dimension_lengths[i]), head.m_start);
+                }
+
+                // Add the assignment to the body
+                std::cout<<ASRUtils::symbol_name(ASR::down_cast<ASR::Var_t>(
head.m_v)->m_v)<<std::endl;
+                flattened_body.push_back(b.Assignment(b.Var(current_scope->r
esolve_symbol(ASRUtils::symbol_name(ASR::down_cast<ASR::Var_t>(head.m_v)->m_v
))),
+                computed_var));
+            }
+            for (size_t i = 0; i < do_loop.n_body; ++i) {
+                flattened_body.push_back(do_loop.m_body[i]);
             }
-            body.push_back(al, b.DoLoop(loop_head.m_v, b.Add(start, b.i32(1)
), end, loop_body, loop_head.m_increment));
+            body.push_back(al, b.DoLoop(loop_head.m_v, b.Add(start, b.i32(1)
), end, flattened_body,
+            loop_head.m_increment)); std::cout<<"ENds"<<std::endl;
+            //  Collapse Ends Here
+            // body.push_back(al, b.DoLoop(loop_head.m_v, b.Add(start, b.i32
(1)), end, loop_body, loop_head.m_increment));
 
             body.push_back(al, ASRUtils::STMT(ASR::make_SubroutineCall_t(al,
 loc, current_scope->get_symbol("gomp_barrier"), nullptr, nullptr, 0, nullptr
)));
```

File: integration_tests/openmp_40.f90 [hunk 2]
```diff
@@ -0,0 +1,6 @@
+program openmp_40
+integer :: i, j
+do concurrent ( i =1:4, j = 1:3 )
+    print *,i
+end do
```
Comment (Pranavchiku) [inline integration_tests/openmp_40.f90:5]: try to add values of  i and test it after loop ends.

**Comment by Pranavchiku** [comment]
Let's not test with GFortran if it is not supported.

File: integration_tests/openmp_40.f90 [hunk 3]
```diff
@@ -0,0 +1,8 @@
+program openmp_40
+integer :: i, j, res=0
+do concurrent ( i =1:4, j = 1:3 ) reduce(+:res)
+    print *,i
+    res=res+i
+end do 
+print *,"res =",res
```
Comment (Pranavchiku) [inline integration_tests/openmp_40.f90:7]:
```suggestion
print *,"res =",res
if ( res /= xx ) error stop
```

Put correct value of `xx` in here

**Review by Pranavchiku — APPROVED** [review]
just apply the two minor changes and this will be good to merge

**Review by certik — APPROVED** [review]

**Pull request merged**